### PR TITLE
feat(mixin): making the actor field configurable

### LINKED ISSUE #68: Make actor field selection configurable
**Describe the bug**

- The actor field in the audit logs created use the id of the user bound in the context and there is no way to override this behaviour
- We can not use the mixin if the user is not present in the context
- deleteAllHard method is not creating any log records so there should be a way to create records for such methods

**To Reproduce**
1. Add mixin to repo
2. Hit the repo with a User in the context
3. The actor saved uses the id property of user

**Expected behavior**

- The property of the user that is saved in the log should be configurable
- It should work without user bound in context
- deleteAllHard should also create log records

**Additional context**
This is required in case we want to save something like `userTenantId` instead of simply the `userId`

### PR DESCRIPTION
**Pull request by yeshamavani**
## Description

1) Added an optional constructor argument actorIdKey that can be injected via application context.
   It can accepts User key of the type string.
   the default value for the actorIdKey will be id as in it will save the userId keeping the current functionality intact.
2) Now hard delete methods are available -- deleteAllHard() and deleteByIdHard()
3) If current user is not bounded to any value in the application context then actorId field can be passed through the options.

Fixes #68

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [x] Performed a self-review of my own code
- [x] npm test passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/mixins/audit.mixin.ts [hunk 1]
```diff
@@ -82,8 +88,11 @@ export function AuditRepositoryMixin<
           data =>
             new AuditLog({
               actedAt: new Date(),
-              // eslint-disable-next-line @typescript-eslint/no-explicit-any
-              actor: (user?.id as any).toString() ?? '0', //NOSONAR
+              actor:
+                options?.actorId ??
```
Comment (akshatdubeysf) [inline src/mixins/audit.mixin.ts:92]: This seems repeated, can't you move this to a function? that can be overridden?

File: src/mixins/audit.mixin.ts [hunk 2]
```diff
@@ -296,8 +317,101 @@ export function AuditRepositoryMixin<
...
+            `Audit failed for data => ${JSON.stringify(auditsJson)}`,
+          );
+          //sonarignore:end
+        });
+      }
+      return deletedCount;
+    }
+
+    async deleteByIdHard(id: ID, options?: AuditOptions): Promise<void> {
+      if (!super.deleteByIdHard) {
```
Comment (akshatdubeysf) [inline src/mixins/audit.mixin.ts:393]: Should throw an error if base class does not implement this method? and Ideally the type should not expose this function at all if the baseClass does not have it

**Review by akshatdubeysf — CHANGES_REQUESTED** [review]

File: src/mixins/audit.mixin.ts (hunk 1 above)
Comment (yeshamavani) [reply src/mixins/audit.mixin.ts]: done

File: src/mixins/audit.mixin.ts [hunk 3]
```diff
@@ -127,8 +128,7 @@ export function AuditRepositoryMixin<
           data =>
             new AuditLog({
               actedAt: new Date(),
-              // eslint-disable-next-line @typescript-eslint/no-explicit-any
-              actor: (user?.id as any).toString() ?? '0', //NOSONAR
+              actor: this.getActor(user, this.actorIdKey, options?.actorId),
```
Comment (akshatdubeysf) [inline src/mixins/audit.mixin.ts:131]: arre why do you need to pass this.actorIdKey, the getActor would have access to it na as that is a member variable?

**Review by akshatdubeysf — CHANGES_REQUESTED** [review]

File: src/mixins/audit.mixin.ts (hunk 3 above)
Comment (yeshamavani) [reply src/mixins/audit.mixin.ts]: rectified

**Review by akshatdubeysf — APPROVED** [review]

**Pull request merged**